service/src/BatchServer.cpp

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright (c) 2015 - 2021, Intel Corporation
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of Intel Corporation nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "BatchServer.hpp"
+
+#include <cstdlib>
+#include <cerrno>
+#include <unistd.h>
+
+#include "geopm/Exception.hpp"
+#include "geopm/PlatformIO.hpp"
+#include "geopm/SharedMemory.hpp"
+#include "geopm/Helper.hpp"
+#include "geopm/PlatformIO.hpp"
+#include "BatchStatus.hpp"
+#include "geopm_debug.hpp"
+
+
+namespace geopm
+{
+    std::unique_ptr<BatchServer>
+    BatchServer::make_unique(int client_pid,
+                             const std::vector<geopm_request_s> &signal_config,
+                             const std::vector<geopm_request_s> &control_config)
+    {
+        return geopm::make_unique<BatchServerImp>(client_pid, signal_config,
+                                                  control_config);
+    }
+    BatchServerImp::BatchServerImp(int client_pid,
+                                   const std::vector<geopm_request_s> &signal_config,
+                                   const std::vector<geopm_request_s> &control_config)
+        : BatchServerImp(client_pid, signal_config, control_config,
+                         platform_io(), nullptr, nullptr, nullptr)
+    {
+        // Fork the server when calling real constructor.
+        auto setup = [this]() {this->create_shmem();};
+        auto run = [this]() {this->run_batch();};
+        m_server_pid = fork_with_setup(setup, run);
```

Comment:
`BatchServerImp::fork_with_setup()` is able to directly call `create_shmem()` and `run_batch`. The `std::function` arguments don't seem necessary.

Reply:
This is to enable testing.  We can substitute other void(void) functions and test that forking them works.

Reply:
A good question.  The answer is that we do not want the lifetime of the shared memory regions to be coupled to the lifetime of the geopmd systemd service process (this should be since server boot).  If we create them after we fork, they should be cleaned up when the forked server ends.